active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,405 @@
+                           "method": "GET",
+                           "apiVersion": "2024-02-01",
+                           "apiAction": "/listClusterAdminCredential",
+                           "query": "?a=1&b=2" // optional
+                           "body": {} // optional
+                        },
+                        {
+                           "type": "JSONPath",
+                           "path": "kubeconfig[0].value"
+                        }
+                     ]
+                  }
+               }
+            },
+            "template": {
+               "extensions": {
+                  "k8s": {
+                     "extension": "Kubernetes",
+                     "version": "1.0.0"
+                     // "config" is removed, it's expected to be provided by properties of the deployment
+                  }
+               }
+            }
+         }
+      }
+   }
+}
+```
+
+### Microsoft.Resources/deployments API changes
+
+This REP extends upon the API changes in REP 0003 by adding a configuration value to the extensions returned from
+a deployment GET. This configuration object will be the same as the configuration object described in the above ARM
+template design changes.
+
+```json
+{
+   // ... deployment contract
+   "extensions": [
+      {
+         "name": "Kubernetes",
+         "alias": "k8s",
+         "version": "1.0.0",
+         "deploymentId": "/subscriptions/.../resourceGroups/.../providers/Microsoft.Resources/deployments/...",
+         "config": {
+            "namespace": {
+               "type": "string",
+               "value": "default"
+            },
+            "kubeConfig": {
+               "type": "directive",
+               "evaluation": [
+                  {
+                     "type": "ArmApiCall",
+                     "resourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
+                     "method": "GET",
+                     "apiVersion": "2024-02-01",
+                     "apiAction": "/listClusterAdminCredential",
+                     "query": "?a=1&b=2" // optional
+                     "body": { } // optional
+                  },
+                  {
+                     "type": "JSONPath",
+                     "path": "kubeconfig[0].value" // decompose?
+                  }
+               ]
+            }
+         }
+      }
+   ]
+}
+```
+
+With the deployments service returning a list of all extensions (across all nested deployments) with their configuration
+directives and returning all extension resources deployed in the template (across all nested deployments), the stacks 
+service can process each extension resource and use the "deploymentId" and "extensionAlias" properties as a composite 
+key to the data in the extensions array.
+
+## Drawbacks
+
+The biggest drawback of this approach is that extension configuration data cannot be sourced from deployment parameters.
```

Comment:
I'm on the fence about whether we should completely block this. Have you considered supporting it, but with a "degraded" experience? I think that could be understandable from a user standpoint.

For example, you can still bring your own credentials for managing an on-prem Kubernetes cluster through Stacks, but it's not going to be resilient to rotations. As long as the user has some mechanism to rotate the credentials, I think it's understandable that the Stack won't be able to delete the resource.

Reply:
We would be able to support parameterization of configuration through ARM resources we can query live, key vault secret, etc, but not raw value secrets because we can't persist those. Stacks will need to be robust to failures when retrieving these secrets and provide clear errors regardless of how the constraints on secret sources are set up.

Reply:
Support for BYO credentials will be there, but only in ways where a fetch is done. No raw values. Current design has details on that.